File: assemblyscript-algorithms/src/algorithms/search/meta-binary-search.ts
export function metaBinarySearch_u64_asm(
  array: StaticArray<u64>,
  target: u64
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}
export function metaBinarySearch_i64_asm(
  array: StaticArray<i64>,
  target: i64
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}
export function metaBinarySearch_f64_asm(
  array: StaticArray<f64>,
  target: f64
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}
export function metaBinarySearch_f32_asm(
  array: StaticArray<f32>,
  target: f32
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}
export function metaBinarySearch_u32_asm(
  array: StaticArray<u32>,
  target: u32
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}
export function metaBinarySearch_i32_asm(
  array: StaticArray<i32>,
  target: i32
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}
export function metaBinarySearch_u16_asm(
  array: StaticArray<u16>,
  target: u16
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}
export function metaBinarySearch_i16_asm(
  array: StaticArray<i16>,
  target: i16
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}
export function metaBinarySearch_u8_asm(
  array: StaticArray<u8>,
  target: u8
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}
export function metaBinarySearch_i8_asm(
  array: StaticArray<i8>,
  target: i8
): i32 {
  let n = array.length;
  let lg = <i32>(NativeMath.log(n - 1) / NativeMath.log(2) + 1);

  let pos = 0;
  for (let i = lg; i >= 0; i--) {
    if (array[pos] == target) return pos;

    let new_pos = pos | (1 << i);

    if (new_pos < n && array[new_pos] <= target) pos = new_pos;
  }

  return array[pos] == target ? pos : -1;
}

File: c-algorithms/src/search/meta-binary-search/meta-binary-search.c
#include <emscripten.h>
#include <math.h>
#include "shared/shared.h"
#include "search/meta-binary-search/meta-binary-search.h";

EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_u64_c(Array1d_u64* array, u64 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_i64_c(Array1d_i64* array, i64 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_f64_c(Array1d_f64* array, f64 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_f32_c(Array1d_f32* array, f32 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_u32_c(Array1d_u32* array, u32 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_i32_c(Array1d_i32* array, i32 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_u16_c(Array1d_u16* array, u16 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_i16_c(Array1d_i16* array, i16 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_u8_c(Array1d_u8* array, u8 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
EMSCRIPTEN_KEEPALIVE
int metaBinarySearch_i8_c(Array1d_i8* array, i8 target) {
    int n = array->length;
    int lg = log2(n-1)+1;
    int pos = 0;
    for (int i = lg ; i >= 0; i--) {
        if (array->items[pos] == target)
            return pos;
 
        // Incrementally construct the
        // index of the target value
        int new_pos = pos | (1 << i);
 
        // find the element in one
        // direction and update position
        if ((new_pos < n) && (array->items[new_pos] <= target))
            pos = new_pos;
    }
    return ((array->items[pos] == target) ? pos : -1);
}
